src/Controls/src/Core/VisualElement/VisualElement.cs

```diff
@@ -1000,6 +1000,14 @@ public bool IsPlatformStateConsistent
 		[EditorBrowsable(EditorBrowsableState.Never)]
 		internal event EventHandler PlatformEnabledChanged;
 
+		[EditorBrowsable(EditorBrowsableState.Never)]
+		internal event EventHandler<LayoutPassEvent> LayoutPassEvent;
+
+		private protected void TriggerLayoutPassEvent(LayoutPassEvent layoutPassEvent)
+		{
+			LayoutPassEvent?.Invoke(this, layoutPassEvent);
+		}
```

Comment:
For the final version of this, we need to be sure `TriggerLayoutPassEvent()` is trimmed away in `Release` builds.

You can do this by implementing a trimmer feature switch:
* https://github.com/dotnet/maui/blob/main/docs/design/FeatureSwitches.md

Then add a new switch like this and return early:
https://github.com/dotnet/maui/blob/b43b974e8960918997243064285265ad1ed52f1e/src/Core/src/RuntimeFeature.cs#L78-L87

If you put `if (!RuntimeFeature.LayoutMetricsSupported) return;` at the top of `TriggerLayoutPassEvent()`.

The trimmer (ILLink) will inline your code to `if (true) return;` and be able to remove a lot of code. Ideally, it would mean all the code in `LayoutPassTimer` disappears, too.

Reply:
using `[System.Diagnostics.Conditional("DEBUG")]`, Like the BCL does for `System.Diagnostics.Debug.WriteLine();` isn't enough?

Reply:
`[System.Diagnostics.Conditional("DEBUG")]` works at compile time. MAUI only ships a single `Release` Microsoft.Maui.dll, etc.

If you used `[Conditional("DEBUG")]` it would only work for local developer builds of MAUI.

Reply:
How bcl ships that then? Or I did look at the wrong implementation?

Reply:
@jonathanpeppers why do you want to trim this?
It's basically doing nothing unless they want to subscribe to the event.
If there are no subscription, event is null, so it won't be invoked.
This is like MeasureInvalidated.
I think it'd be valuable to have this available in Release mode if someone wants to measure actual timings.
Can we make it so it gets trimmed on the final app only when not used?
Is that what you're suggestions?

Reply:
Also, how can I verify everything has been trimmed correctly?

Reply:
> It's basically doing nothing unless they want to subscribe to the event.

If it is really doing nothing, maybe rerunning one of the benchmarks would just show if it impacts anything. I mainly want to avoid what Xamarin.Forms did:

* https://github.com/xamarin/Xamarin.Forms/pull/8963

Some internal profiling was added, but then it slowed down *all apps* in `Release` mode!

> how can I verify everything has been trimmed correctly?

Build a project template with your changes `.\dotnet-local.cmd new maui` && `.\dotnet-local.cmd build -c Release -f net9.0-android`.

Then you could open files like `obj\Release\net10.0-android\android-arm64\linked\Microsoft.Maui*.dll` in [ILSpy](https://github.com/icsharpcode/ILSpy) or your favorite decompiler.

Reply:
@jonathanpeppers I tried everything you said, but both class and methods are still there.
I've pushed the changes so that you can verify.
![image](https://github.com/user-attachments/assets/79c78143-41a0-4e19-b170-40ea1770c1a3)